python/lib/sift_client/_internal/low_level_wrappers/channels.py

```diff
@@ -0,0 +1,454 @@
+        not_cached_channels = []
+        for id in channel_ids:
+            if self.channel_cache.channels.get(id):
+                cached_channels.append(id)
+            else:
+                not_cached_channels.append(id)
+        return cached_channels, not_cached_channels
+
+    def _check_cache(
+        self,
+        *,
+        channel_id: str,
+        start_time: datetime,
+        end_time: datetime,
+        run_id: str | None = None,
+    ) -> Tuple[pd.DataFrame | None, datetime | None, datetime | None]:
+        """
+        Check if the data for a channel during a run is cached and return how to query remaining data if so.
+
+        There are a variety of requested start/end time vs cached start/end time cases to consider.
+        Below diagram represents time aligned ranges for each case:
+
+        Cache interval:               |-------------------------------|
+        Case 1:                         |---------------------------|
+        Case 2:                              |--------------------------------|
+        Case 3:                                                           |----------|
+        Case 4:                 |--------------------------------|
+        Case 5:         |------| or |-----------------------------------------|
+
+        Returns:
+            A tuple of (data, start_time, end_time)
+            where data is a pandas dataframe and start and end times are what should be used for the next call based on what is not covered by the cached data.
+        """
+        cached_data = self.channel_cache.channels.get(channel_id)
+        ret_start_time = start_time
+        ret_end_time = end_time
+        ret_data = None
+        if cached_data:
+            start_time_cached = cached_data.start_time
+            end_time_cached = cached_data.end_time
+            ret_data = cached_data.data
+            # Filter data to desiredtime range
+            ret_data = ret_data[start_time:end_time]  # type: ignore # mypy doesn't understand pandas that well seemingly
+
+            if start_time_cached <= start_time:
+                if start_time < end_time_cached:
+                    if end_time <= end_time_cached:
+                        # Case 1
+                        ret_start_time = None  # type: ignore
+                        ret_end_time = None  # type: ignore
+                    else:
+                        # Case 2
+                        ret_start_time = end_time_cached
+                        ret_end_time = end_time
+                else:
+                    # Case 3
+                    return (None, start_time, end_time)
+            else:
+                if start_time_cached < end_time and end_time <= end_time_cached:
+                    # Case 4
+                    ret_start_time = start_time
+                    ret_end_time = start_time_cached
+                else:
+                    # Case 5
+                    return (None, start_time, end_time)
```

Comment:
Maybe a future optimization could be, for the second example in the drawing for Case 5, to just query for the bits outside the beginning/end of the cached window? Luck implemented this behavior in the Grafana plugin so you could get his advice on this.

Reply:
Yeah, I think it'll depend on the situation a bit -- if a user is only requesting a small time window it might be the case that one contiguous call is better, but if they're requesting large chunks of data over large time spans, 2 separate smaller calls may be the better option. maybe something we can wait and see how customers are using this API most often to see what the best approach will be.